### DISCUSSION TITLE
DID registry on Ethereum

### DISCUSSION THREAD

**Discussion by simibac** [decentralized-identity/veramo#360] [Q&A] [answered]
Two questions regarding the DID registration on any of the public testnets?
1. What contract is used to store the DIDs? If my analysis from the dependencies is correct it should be [this contract here.](https://github.com/uport-project/ethr-did-registry/blob/develop/contracts/EthereumDIDRegistry.sol).
2. How do you handle the necessary gas costs that occur during the registration? Do you use the [GSN](https://docs.opengsn.org/#the-problem) for covering the gas cost? If so, how is this configurable?

**Comment by mirceanis** [comment, marked as answer]
Great questions!

1. Yes, that is the contract that is used when `did:ethr` is used.
2. The `did:ethr` protocol does not require DIDs to be registered when they are created.

## Background
The expected lifecycle of a `did:ethr` is that a keypair is created locally and then the public key is used to form the DID:
`generate keyPair` -> ``did:ethr:${publicKey}``
This fresh DID is a valid and resolvable `did:ethr`, and when it is resolved it produces a DID document containing the public key.
Up until this point it is very similar to `did:key`, except for the encoding of the publicKey which in the case of did:ethr is HEX.

The difference comes when you need to update the DID document.
Only then does one need to interact with the ethereum network and use gas.

From the ethr-did-registry contract's perspective, gas can be supplied either by directly funding the address corresponding to the keypair (classical ethereum external account interaction) or by using meta-transactions.

The `ethr-did-registry` contract supports meta transactions since inception, although I don't know if it is compatible with GSN.
GSN seems to be a newer than the already deployed ethr-did-registry. If you do try out using the GSN with the ethr-did-registry then it would make a great entry for the [show and tell](https://github.com/uport-project/veramo/discussions/categories/show-and-tell) section of the discussions 😉

## How it works in veramo now

This repository contains a `did-provider-ethr` plugin that can create and manage ethr DIDs.
Creating is free, but management involves gas.
The current implementation is a very simple one that assumes the address of the did:ethr is already funded.
(improvements here are welcome, by the way)

## Notes
The `did:ethr` protocol differentiates between ethereum networks:
* `did:ethr:${publicKey}` -> mainnet
* `did:ethr:rinkeby:${publicKey}` -> rinkeby (using the network name)
* `did:ethr:0x4:${publicKey}` -> also rinkeby (using chainID)

This means that the same keypair can control multiple DIDs; each resolvable using a different ethereum network.

**Reply by nayse** [reply]
Hi. Does this process of creating the did:ethr can done inside a Besu enviroment? Can I register issuers, schemas, and definitions from verifiable credentials using the smart contract in the blockchain?
